dali/imgcodec/decoders/nvjpeg.cc

```diff
@@ -0,0 +1,223 @@
+// Copyright (c) 2019-2022, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <map>
+#include <string>
+#include <utility>
+#include "dali/core/device_guard.h"
+#include "dali/imgcodec/decoders/nvjpeg.h"
+#include "dali/imgcodec/decoders/nvjpeg/nvjpeg_helper.h"
+#include "dali/imgcodec/decoders/nvjpeg/nvjpeg_memory.h"
+#include "dali/imgcodec/decoders/nvjpeg/permute_layout.h"
+
+namespace dali {
+namespace imgcodec {
+
+NvJpegDecoderInstance::
+NvJpegDecoderInstance(int device_id, const std::map<std::string, any> &params)
+: BatchParallelDecoderImpl(device_id, params)
+, device_allocator_(nvjpeg_memory::GetDeviceAllocator())
+, pinned_allocator_(nvjpeg_memory::GetPinnedAllocator()) {
+  DeviceGuard dg(device_id_);
+  CUDA_CALL(nvjpegCreateSimple(&nvjpeg_handle_));
+
+  SetParams(params);
+
+  any num_threads_any = GetParam("nvjpeg_num_threads");
```

Comment:
This won't work - note that your implementation of `GetParam` doesn't even implement this argument. Please add appropriate member - and when you have it, you can use it directly, without going through GetParam here.